controller/api/destination/server.go

```diff
@@ -209,23 +211,33 @@ func (s *server) GetProfile(dest *pb.GetDestination, stream pb.Destination_GetPr
 			if err != nil {
 				return err
 			}
-
-			// The IP may or may not map to a pod (pod argument can be nil). If
-			// pod is not nil we will return a single endpoint in the
-			// DestinationProfile response, otherwise we return a default
-			// profile response.
-			err = s.sendEndpointProfile(stream, pod, port)
-			if err != nil {
-				log.Debugf("Failed to send profile response for endpoint %s:%d: %v", ip.String(), port, err)
-				return err
+			if pod == nil {
+				translator := newEndpointProfileTranslator(nil, port, nil, nil, stream, s.log)
+				translator.UpdateProtocol(false)
+			} else {
+				address, err := s.createAddress(pod, port)
+				if err != nil {
+					s.log.Errorf("failed to create address: %s", err)
+				}
+				opaquePorts, err := getAnnotatedOpaquePorts(pod, s.defaultOpaquePorts)
+				if err != nil {
+					s.log.Errorf("failed to get opaque ports for pod %s/%s: %s", pod.Namespace, pod.Name, err)
+				}
+				endpoint, err := s.createEndpoint(address, pod, opaquePorts)
+				if err != nil {
+					return err
+				}
+				translator := newEndpointProfileTranslator(pod, port, endpoint, opaquePorts, stream, s.log)
```

Comment:
if endpoint is in opaquePorts, the opaqueness will always be true, regardless of any servers, right?  so we could check if endpoint is in opaquePorts here and if it is, then we can just send one update and then skip subscribing to services or creating a translator.  this also means that we wouldn't need to pass opaquePorts into the endpointProfileTranslator and could simplify the logic in there